### ISSUE TITLE
Create an interface for creating a mutable copy of immutable entities

### ISSUE THREAD

**Issue by vestrel00** [vestrel00/contacts-android#115]
In attempts to solve #109, I hit a roadblock due to the fact that `toMutable` functions of all entities are not abstracted away by a facade. This means that generic `ImmutableData` types cannot be converted to the generic counterpart `MutableData` type. This makes the following things impossible to do without doing unsafe type checking and casting,

```kotlin
val immutableData: ImmutableData
val mutableData: MutableData = immutableData.toMutable???

fun <K : ImmutableData, V : MutableData> K.toMutable???: V

fun <K : ImmutableData, V : MutableData> List<K>.toMutable???: List<V>
``` 

There are some immutable entities (e.g. GroupMembership and Photo) that does not have a mutable counterpart. That also needs to be taken into consideration. For those that have a mutable version, introduce something like,

```kotlin
ImmutableEntity<M : MutableEntity> {
    fun toMutableEntity(): M
}
```

Or something more concise like,

```kotlin
ImmutableEntity<M : MutableEntity> {
    fun mutableCopy(): M
}
```

## Checklist

- [x] Complete the refactor
- [x] Look through a bunch of Contact/RawContact and MutableContact/MutableRawContact extension functions in the util package. Check if the output of this issue enables consolidation of "duplicate code". Can we just declare one function via ContactEntity/RawContactEntity? If so, create a separate issue to perform this refactor.
- [x] Update mentions of "toMutable" everywhere
- [x] Update mentions of ".Type" everywhere
- [x] Update relevant documentation in code
- [x] Update relevant documentation in howto pages
- [x] Update relevant documentation in README
- [x] Update relevant documentation in DEV_NOTES
- [ ] Verify that there are no regressions for all API functions!

## Notes

Sealed interfaces will probably come in handy here. So this should be worked on after #121 and #123

**Comment by vestrel00** [comment]
This is a part of [v0.1.9 Release Checklist](https://github.com/vestrel00/contacts-android/discussions/132).
